virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,317 @@
+            self.topofcanopy_radiation,
+            data["canopy_absorption"],
+            self.longwave_radiation,
+        )
+
+
+# helper functions
+def calculate_atmospheric_transmissivity(
+    elevation: NDArray[np.float32],
+    sunshine_fraction: NDArray[np.float32] = np.array(1.0, dtype=np.float32),
+    cloudy_transmissivity: float = CLOUDY_TRANSMISSIVITY,
+    transmissivity_coefficient: float = TRANSMISSIVITY_COEFFICIENT,
+    beer_regression: float = BEER_REGRESSION,
+) -> NDArray[np.float32]:
+    """Calculate atmospheric transmissivity (tau).
+
+    Args:
+        elevation: elevation above sea level, [m]
+        sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+            and 1 (cloud free sky), default = 1
+        cloudy_transmissivity: cloudy transmittivity :cite:p:`Linacre1968`, default set
+            in config
+        transmissivity_coefficient: angular coefficient of transmittivity
+            :cite:p:`Linacre1968`, default set in config
+        beer_regression: parameter in equation for atmospheric transmissivity based on
+            regression of Beer's radiation extinction function :cite:p:`Allen1996`,
+            default set in config
+
+    Returns:
+        atmospheric transmissivity, unitless
+    """
+
+    # check sunshine fraction between 0 and 1
+    if 0 > np.any(sunshine_fraction) > 1:
+        to_raise = ValueError(
+            "The fraction of sunshine hours needs to be between 0 and 1!"
+        )
+        LOGGER.critical(to_raise)
+        raise to_raise
+
+    # calculate transmissivity (tau), unitless
+    tau_o = cloudy_transmissivity + transmissivity_coefficient * sunshine_fraction
+    return tau_o * (1.0 + beer_regression * elevation)
+
+
+def calculate_ppfd(
+    shortwave_in: NDArray[np.float32],
+    elevation: NDArray[np.float32],
+    sunshine_fraction: NDArray[np.float32] = np.array(1.0, dtype=np.float32),
+    albedo_vis: NDArray[np.float32] = np.array(0.03, dtype=np.float32),
+    flux_to_energy: float = FLUX_TO_ENERGY,
+) -> NDArray[np.float32]:
+    """Calculate top of canopy photosynthetic photon flux density, [mol m-2].
+
+    Args:
+        shortwave_in: downward shortwave radiation, [J m-2]
+        elevation: elevation above sea level, [m]
+        sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+            and 1 (cloud free sky), default = 1
+        albedo_vis: visible light albedo, default = 0.03
+        flux_to_energy: flux to energy conversion factor, [umol J-1],default from config
+
+    Returns:
+        photosynthetic photon flux density, [mol m-2]
+
+    Reference: :cite:t:`Davis2017`
+    """
+
+    # check if sunshine fraction is in data object, if not set to 1
+    if sunshine_fraction.size > 1:
+        sunshine_fraction = data["sunshine_fraction"]
+    else:
+        sunshine_fraction = np.ones(len(elevation), dtype=np.float32)
+        LOGGER.info("The sunshine fraction is set to default = 1")
```

Comment:
This `if`-statement means that *whatever* value for `sunshine_fraction` is passed into the function, it will be discarded and replaced either by ones or `data["sunshine_fraction"]`. I don't think this is what you want.

Reply:
yes, that's true. We changed this to check if the input is provided and if not, set to default. This does not check if any values are missing but this should be picked up by earlier checks?